### DISCUSSION TITLE
App presets (group of extensions) to ease generating Quarkus apps

### DISCUSSION THREAD

**Discussion by ia3andy** [quarkusio/quarkus#25757] [Community]
Starting a discussion about https://github.com/quarkusio/quarkus/issues/24657

**Comment by ia3andy** [comment]
ping @quarkusio/devtools

**Comment by aloubyansky** [comment]
There could be different approaches here. The simplest one would probably be creating named sets of extension GAs (actually GACTs) and adding them to the platform JSON descriptors. Dev tools could then expose them to the users allowing to be selected, etc.

I wanted to throw out a bit more sophisticated idea I explored once for consideration. Instead of configuring very precise (implementation-based) presets, we could consider more general technology “packs” (allowing alternative implementations, e.g. JDBC drivers, REST, etc). It’s also somewhat related to issues in [1] (where referencing specific extensions in codestart caused an incompatibility issue) and general extension set compatibility checking.
I was wondering about leveraging the capability mechanism https://quarkus.io/guides/capabilities for this. Note, that capabilities are already included in the extension metadata and, given an extension catalog, we know which extension provides which capabilities and which capabilities it requires (if any).

E.g. we know that `quarkus-resteasy` and `quarkus-resteasy-reactive` provide capability `io.quarkus.rest`. And so they are alternative implementations of the REST capability and also can’t be used in the same project.
We could imagine that the JDBC driver extensions would provide capabilities with `io.quarkus.jdbc.` prefix. E.g. `io.quarkus.jdbc.psql`, `io.quarkus.jdbc.mysql`, etc.

An extension pack then could be expressed as a set of capabilities that must be satisfied. It would still have some user-friendly name and be recorded in the platform metadata. E.g. an extension REST+CRUD pack could be configured as requiring capabilities:
`io.quarkus.rest`
`io.quarkus.jdbc.*`

Again, given an extension catalog, we know which extensions provide these capabilities. We could either highlight to the user (or offer dropdowns, depending on the UI we come up with) the extensions to choose from to fulfill the chosen pack requirements or pre-select specific implementations. To be able to pre-select specific implementations though, we’d need to specify the default/preferred implementation choice of a given capability, unless a user has already selected one. This would have to be recorded in the platform metadata too.

An advantage of this indirection is that if/when we switch from one preferred impl of a tech to another one (e.g. resteasy -> resteasy-reactive) we wouldn't have to go an update GAs everywhere. We'd simply update the preferred impl of a given capability in the metadata. And I think a similar approach could help with issues in [1].

Whatever the user chooses we’d still perform the capability requirements check for the selected extensions to make sure there are no conflicts and the requirements are satisfied.

[1] https://github.com/quarkusio/quarkus/issues/25384

**Reply by maxandersen** [reply]
What does jdbc.* mean? Only one match or any amount of jdbc ?

**Reply by aloubyansky** [reply]
> What does jdbc.* mean? Only one match or any amount of jdbc ?

That at least one capability with `io.quarkus.jdbc.` prefix is present.

**Reply by gastaldi** [reply]
I like the qualified name constraint idea on capabiltities, eg. 
- `io.quarkus.rest.classic` and `io.quarkus.rest.reactive` matched when `io.quarkus.rest` is requested
- `io.quarkus.jdbc.postgres`, etc matching `io.quarkus.jdbc` is requested

**Reply by aloubyansky** [reply]
> I like the qualified name constraint idea on capabiltities, eg.
> 
> * `io.quarkus.rest.classic` and `io.quarkus.rest.reactive` matched when `io.quarkus.rest` is requested

Today they both provide `io.quarkus.rest` but `quarkus-resteasy` also provides `io.quarkus.resteasy` while `quarkus-resteasy-reactive` provides `io.quarkus.resteasy.reactive`. So the capabilities are different but have the same prefix. These are details though that we could improve if necessary.

**Reply by maxandersen** [reply]
I like the idea of using capability ids as looser way to define it but we would also need to state explicit extensions but shouldn't be a problem. 

In fact if we do this we could allow this on quarkus create and mvn plugin too in part of its "search" for matches.

**Reply by aloubyansky** [reply]
@maxandersen could you clarify what you mean by the need for stating explicit extensions? In the named "pack" definition in addition to the capability names we'd need to list extension GAs?
I didn't mean the capabilities would have to be exposed to the user actually. Although they could be as an alternative way to search/select extensions. I probably wouldn't do it right from the start though.

**Reply by maxandersen** [reply]
I'm referring to extensions that aren't expressed as capabilities (my understanding is that not all does this).

Even if they do users (or quarkus tooling) would need to pin some extensions to be able to resolve the right set.

**Reply by aloubyansky** [reply]
It's easy to add a capability though. I think it'd be easier that way than supporting two alternative ways of defining these presets.

**Reply by ia3andy** [reply]
As appealing as it seems, I don't see any real value added in having those group of capabilities. I think having a "capabilities" view in the tooling might be more user-friendly, in this view you would pick capabilities then implementation.

Also if we want to bind codestarts to those presets, the logic depends on the chosen capabilities implementations. 

That said, for capabilities which are not code dependent (i.e jdbc), then we could have the preset reference a capability. Do we have a lot of those?

**Reply by aloubyansky** [reply]
I don't think we should link codestarts to capabilities. Codestarts should remain to be linked to extensions, IMO. How would you model/configure presets, @ia3andy?

**Reply by maxandersen** [reply]
Not following how things would work without being able to use capabilities AND specific extensions ? Majority of extensions aren't defined as capabilities or am I missing something ?

**Reply by ia3andy** [reply]
If we introduce the "capabilities" view that I described above, maybe it would solve the initial problem (and remove the need for preset). I could maybe come up with a POC?

> I don't think we should link codestarts to capabilities. Codestarts should remain to be linked to extensions, IMO. How would you model/configure presets, @ia3andy?

If we decide to go for presets, I didn't mean to link codestarts to capabilities but to presets. Of course, we first need to see if it makes sense, if some of the presets would benefit from having starter code. Note that I created an issue https://github.com/quarkusio/quarkus/issues/25862 about adding the notion of variant to allow having different codestarts for a single extension (or preset).

**Reply by justinjohnson02** [reply]
I will add $0.02 that the current extensions model is very nice and easy to "consume" when building or putting together an app or stack. One of the things that makes Spring "massive" (IMO) is the plethora of seemingly large auto-wired "bundles".

**Reply by aloubyansky** [reply]
Thanks @justinjohnson02 

@ia3andy I briefly looked into https://github.com/quarkusio/quarkus/issues/25384 and it looks like the basic mechanism and config would pretty much be the same. Presets then would be simply a grouping on top of that.

However, talking about #25384 and other recently open issues about adding extra platform BOMs, etc, it will require some refactoring. Basically, there is a chicken-egg issue. If we allow a codestart to reference another platform extension for which we need to import another BOM or a capability (instead of a direct extension GAV) that would be resolved to an extension, we end up with codestarts as another input source of extensions to be installed. And so, we resolve codestarts from the extensions and then from the codestarts we may need to install other extensions (that may require new BOM imports).

BTW, couldn't we fix #25384 by extracting the spring-web codestart to its own artifact and referencing it from the spring-web extension? Yes, it'll result more codestart artifacts that need to be resolved. OTOH, it seems safer from compatibility perspective.

**Reply by ia3andy** [reply]
@aloubyansky we could put the resteasy&reactive/spring-web codestarts with the other core extension codestarts, it would have the same result right?

**Reply by aloubyansky** [reply]
As long as it's included in the codestarts bundled with tools, it'll be fragile like this.

**Reply by ia3andy** [reply]
@aloubyansky the core extensions codestarts are platform dependent (the base codestarts are bundled)

**Reply by aloubyansky** [reply]
Then it should be enough.

**Reply by ia3andy** [reply]
It will break backward compatibility and it may also break forward compat for previous CLI versions

**Reply by aloubyansky** [reply]
Is base codestart always preferred over a non-base one?

**Reply by ia3andy** [reply]
I need to check but I think yes

**Reply by aloubyansky** [reply]
Could you describe how it would break then?

**Reply by ia3andy** [reply]
Ok, I verified base codestarts override extensions codestarts.

Well it depends,
- When using old CLI with a recent version, you will get the extensions codestarts embeded in that CLI (not broken but not what you would expect)
- When using a recent CLI with an old version, then they will not contain those codestarts, you won't have them.

**Reply by maxandersen** [reply]
"When using old CLI with a recent version, you will get the extensions codestarts embeded in that CLI (not broken but not what you would expect)"

not following - why would that not be what you expect ? 

"When using a recent CLI with an old version, then they will not contain those codestarts, you won't have them."

makes sense to me - groupings is per platform so old platforms won't have them.

**Reply by ia3andy** [reply]
> "When using a recent CLI with an old version, then they will not contain those codestarts, you won't have them."
> 
> makes sense to me - groupings is per platform so old platforms won't have them.

I am talking about RESTEasy C and R (which are default) and Spring Web, not having them is a problem IMO?

**Reply by maxandersen** [reply]
Not following what you mean about "not having them" - what is "them" in that sentence?

I talked about old versions of platform not having groupings. Which I don't see why that is a problem. Can you outline where you think that is an issue ?

**Comment by maxandersen** [comment]
"It's easy to add a capability though. I think it'd be easier that way than supporting two alternative ways of defining these presets."

if we do that it means we cannot add extensions that don't have capabilites and if we force it on all now everyone has to think really hard about what kind of capability they provide...and why would that not just end up being the exact extension id name for most ?

I still think you would just use capablities when you actually share api/functionallity with multiple extensions ... extensions group should be able to be defined via explicit extensions + capability (with default choice if no other way to decide which of many). Otherwise its too narrow usage.

**Reply by aloubyansky** [reply]
I am not sure I like mixing capabilities and specific implications (I.e. GAV). Specific implications provide capabilities which we would have to take into account anyway. Also we are thinking of capabilities for a reason here, which is to **avoid** referencing specific GAVs from codestarts. Given that presets would most probably be limited to particular extensions, imo it wouldn't be a big deal to require a capability for presets.
